pylearn2/utils/iteration.py

```diff
@@ -639,6 +640,111 @@ def __next__(self):
+    the same length.
+    Within one minibatch all sequences will have the same lenght, so it
+    might return minibatches with different sizes depending on the
+    distribution of the lengths of sequences in the data.
+
+    Notes
+    -----
+    Returns lists of indices (`fancy = True`).
+
+    Parameters
+    ----------
+    sequence_data : list of lists or ndarray of objects (ndarrays)
+        The sequential data used to determine indices within the dataset such
+        that within a minibatch all sequences will have same lengths.
+
+    See :py:class:`SubsetIterator` for detailed constructor parameter
+    and attribute documentation.
+    """
+
+    def __init__(self, sequence_data, batch_size, num_batches=None, rng=None):
+        self._rng = make_np_rng(rng, which_method=["random_integers",
+                                                   "shuffle"])
+
+        if batch_size is None:
+            raise ValueError("batch_size cannot be None for random uniform "
+                             "iteration")
+        if num_batches is not None:
+            raise ValueError("EvenSequencesSubsetIterator doesn't support"
+                             " fixed number of batches")
+        if isinstance(sequence_data, list):
+            self._dataset_size = len(sequence_data)
+        elif isinstance(sequence_data, np.ndarray):
+            self._dataset_size = sequence_data.shape[0]
+        else:
+            raise ValueError("sequence_data must be of type list or np.ndarray")
+        self._sequence_data = sequence_data
+        self._batch_size = batch_size
+        self.prepare()
+        self.reset()
+
+    def prepare(self):
+        # find the unique lengths
+        self.lengths = [len(s) for s in self._sequence_data]
+        self.len_unique = np.unique(self.lengths)
+
+        # indices of unique lengths
+        self.len_indices = dict()
+        self.len_counts = dict()
+        for ll in self.len_unique:
+            self.len_indices[ll] = np.where(self.lengths == ll)[0]
+            self.len_counts[ll] = len(self.len_indices[ll])
+
+    def reset(self):
+        self.len_curr_counts = copy.copy(self.len_counts)
+        self.len_unique = self._rng.permutation(self.len_unique)
+        self.len_indices_pos = dict()
+        for ll in self.len_unique:
+            self.len_indices_pos[ll] = 0
+            self.len_indices[ll] = self._rng.permutation(self.len_indices[ll])
+        self.len_idx = -1
+
+    @wraps(SubsetIterator.next)
+    def next(self):
+        # randomly choose the length
+        count = 0
+        while True:
+            self.len_idx = np.mod(self.len_idx+1, len(self.len_unique))
+            if self.len_curr_counts[self.len_unique[self.len_idx]] > 0:
+                break
+            count += 1
+            if count >= len(self.len_unique):
+                break
+        if count >= len(self.len_unique):
+            self.reset()
+            raise StopIteration()
```

Comment:
This loop deserves a bit more comments. Also you can significantly improve readability by refactoring the checks for "is it time to reset?" out of the loop. For instance, first sum the values in `self.len_curr_counts` and if the sum is 0, it's time to reset and raise the exception. Otherwise, do the loop to choose the length.